actions/actions.py

```diff
@@ -0,0 +1,180 @@
+    with tarfile.open(backup_path, 'r:gz') as tar:
+        try:
+            tar.extractall(easyrsa_directory)
+        except KeyError:
+            raise RuntimeError('Backup file "{}" does not contain expected '
+                               'files'.format(backup_name))
+
+    # Set these values on the leadership data.
+    cert_dir = os.path.join(pki_dir, 'issued')
+    key_dir = os.path.join(pki_dir, 'private')
+    ca_cert = os.path.join(pki_dir, 'ca.crt')
+    ca_key = os.path.join(key_dir, 'ca.key')
+    serial_file = os.path.join(pki_dir, 'serial')
+
+    with open(ca_cert, 'r') as stream:
+        print("Writing ca")
+        leader_set({
+            'certificate_authority': stream.read()})
+
+    with open(ca_key, 'r') as stream:
+        print("Writing key")
+        leader_set({
+            'certificate_authority_key': stream.read()})
+
+    with open(serial_file, 'r') as stream:
+        print("Writing serial")
+        leader_set({
+            'certificate_authority_serial': stream.read()})
+
+    ca_cert = leader_get('certificate_authority')
+    print(ca_cert)
+    tls = endpoint_from_name('client')
+    tls.set_ca(ca_cert)
+    for client in tls.all_requests:
+        try:
+            cert_file = os.path.join(cert_dir,
+                                     "{}.crt".format(client.common_name))
+            key_file = os.path.join(key_dir,
+                                    "{}.key".format(client.common_name))
+            with open(cert_file, 'r') as file:
+                cert = file.read()
+                print('writing cert to '
+                      '{}:\n{}'.format(client.common_name, cert))
+            with open(key_file, 'r') as file:
+                key = file.read()
+            client.set_cert(cert, key)
+
+        except FileNotFoundError:
+            print("Cert not found for {}".format(client.common_name))
+
+
+def list_backups() -> None:
+    file_list: List[str] = []
+
+    try:
+        file_list = os.listdir(PKI_BACKUP)
+    except FileNotFoundError:
+        pass
+
+    if file_list:
+        message = 'Available backup files:'
+        for file in file_list:
+            message += '\n{}'.format(file)
+    else:
+        message = 'There are no available backup files.'
+
+    function_set({'message': message})
+
+
+def delete_backup() -> None:
+    backup_name: str = function_get('name')
+    delete_all: bool = function_get('all')
+
+    if not delete_all:
+        if backup_name is None:
+            raise RuntimeError('Parameter \'name\' is required if parameter '
+                               '\'all\' is False.')
+        delete_file = os.path.join(PKI_BACKUP, backup_name)
+        try:
+            os.remove(delete_file)
+        except FileNotFoundError:
+            raise RuntimeError('Backup file "{}" does not '
+                               'exist'.format(backup_name))
+    else:
+        shutil.rmtree(PKI_BACKUP)
+        os.mkdir(PKI_BACKUP)
```

Comment:
Suggestion:
```python
if delete_all:
    shutil.rmtree(PKI_BACKUP)
    os.mkdir(PKI_BACKUP)
    # I think this step misses the os.chown call done
    # during the backup action (maybe refactor that part
    # so both functions can (re)create the PKI_BACKUP folder
    return
# then, the if backup_name is None, etc. without indentation
```

Also, shutil.rmtree is called without checking if that path is a directory? Is this intended? The backup action checked `isdir`.

Reply:
> I think this step misses the os.chown call done

Yeah and it's even unnecessary to recreate the dir here..since that is handled in the `backup` action. All that is necessary here is the `rmtree` call

> then, the if backup_name is None, etc. without indentation

I don't know, I would rather have an `if-else` with single return at then end of the function.

> Also, shutil.rmtree is called without checking if that path is a directory? Is this intended? The backup action checked isdir

hmm...I mean we can check for it...but what if it isn't dir? should the `delete` action fail? The `backup` action checks for it because it makes it impossible to proceed with the action, however here nothing stops us from clearing the designated backup space.

Reply:
Protection against directory traversal was added. I'm gonna mark conversation as `Resolved`